File: server/src/controllers/PlaidController.js
import { PlaidService } from '../services/PlaidService.js';
import { AccountMapper, PlaidItemMapper } from '../mappers/index.js';

export class PlaidController {
  constructor() {
    this.plaidService = new PlaidService();
  }

  /**
   * Create link token for Plaid Link
   * @param {Object} req - Express request
   * @param {Object} res - Express response
   */
  createLinkToken = async (req, res) => {
    try {
      const userId = req.user?.id || 'user-id'; // TODO: Get from auth middleware
      const linkToken = await this.plaidService.createLinkToken(userId);

      res.json({
        success: true,
        data: {
          linkToken: linkToken.link_token,
          expiration: linkToken.expiration,
          requestId: linkToken.request_id
        }
      });
    } catch (error) {
      console.error('Error creating link token:', error);
      res.status(500).json({
        success: false,
        error: 'Failed to create link token'
      });
    }
  };

  /**
   * Exchange public token for access token
   * @param {Object} req - Express request
   * @param {Object} res - Express response
   */
  exchangePublicToken = async (req, res) => {
    try {
      const { publicToken, metadata } = req.body; // Note: camelCase from client
      const userId = req.user?.id || 'user-id'; // TODO: Get from auth middleware

      if (!publicToken) {
        return res.status(400).json({
          success: false,
          error: 'Public token is required'
        });
      }

      const result = await this.plaidService.exchangePublicToken(
        publicToken, 
        metadata, 
        userId
      );

      // Start initial transaction sync in background
      setTimeout(() => {
        this.plaidService.getInitialTransactions(
          result.plaidItem.access_token, 
          userId
        ).catch(console.error);
      }, 2000);

      res.json({
        success: true,
        data: {
          itemId: result.plaidItem.item_id,
          accountsLinked: result.accountCount,
          institution: metadata?.institution?.name,
          accounts: AccountMapper.toDTOArray(result.accounts || [])
        }
      });
    } catch (error) {
      console.error('Error exchanging public token:', error);
      res.status(500).json({
        success: false,
        error: 'Failed to exchange token'
      });
    }
  };

  /**
   * Get user accounts
   * @param {Object} req - Express request
   * @param {Object} res - Express response
   */
  getAccounts = async (req, res) => {
    try {
      const userId = req.params.userId || req.user?.id || 'user-id';
      const accounts = await this.plaidService.getAccountsByUser(userId);

      const accountDTOs = AccountMapper.toDTOArray(accounts);

      res.json({
        success: true,
        data: accountDTOs,
        meta: {
          totalAccounts: accountDTOs.length,
          activeAccounts: accountDTOs.filter(acc => acc.isActive).length
        }
      });
    } catch (error) {
      console.error('Error fetching accounts:', error);
      res.status(500).json({
        success: false,
        error: 'Failed to fetch accounts'
      });
    }
  };

  /**
   * Sync transactions for user
   * @param {Object} req - Express request
   * @param {Object} res - Express response
   */
  syncTransactions = async (req, res) => {
    try {
      const { userId } = req.body; // Note: camelCase from client
      const finalUserId = userId || req.user?.id || 'user-id';

      if (!finalUserId) {
        return res.status(400).json({
          success: false,
          error: 'User ID is required'
        });
      }

      const newTransactions = await this.plaidService.syncUserTransactions(finalUserId);
      const transactionDTOs = TransactionMapper.toDTOArray(newTransactions);

      res.json({
        success: true,
        data: {
          transactionsSynced: transactionDTOs.length,
          transactions: transactionDTOs
        }
      });
    } catch (error) {
      console.error('Error syncing transactions:', error);
      res.status(500).json({
        success: false,
        error: 'Failed to sync transactions'
      });
    }
  };

  /**
   * Create sandbox connection (for testing)
   * @param {Object} req - Express request
   * @param {Object} res - Express response
   */
  sandboxAutoConnect = async (req, res) => {
    try {
      const userId = req.user?.id || 'custom_dnc_user'; // TODO: Get from auth middleware

      const result = await this.plaidService.createSandboxConnection(userId);

      res.json({
        success: true,
        data: {
          itemId: result.plaidItem.item_id,
          accountsLinked: result.accountCount,
          institution: 'Platypus Bank (Sandbox)',
          message: 'Sandbox connection created successfully',
          accounts: AccountMapper.toDTOArray(result.accounts || [])
        }
      });
    } catch (error) {
      console.error('Sandbox auto connect failed:', error);
      res.status(500).json({
        success: false,
        error: 'Sandbox connection failed'
      });
    }
  };
}

File: server/src/main/java/com/fintrack/plaid/PlaidController.java
package com.fintrack.plaid;

import com.fintrack.common.ApiResponse;
import com.fintrack.security.AuthUtils;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/plaid")
public class PlaidController {
  private final PlaidService plaidService;

  public PlaidController(PlaidService plaidService) {
    this.plaidService = plaidService;
  }

  @PostMapping("/create_link_token")
  public ResponseEntity<ApiResponse<Map<String, String>>> createLinkToken(Authentication authentication) {
    try {
      String userId = AuthUtils.requireUserId(authentication);
      String linkToken = plaidService.createLinkToken(userId);
      return ResponseEntity.ok(ApiResponse.success(Map.of("linkToken", linkToken)));
    } catch (Exception e) {
      return ResponseEntity.badRequest().body(ApiResponse.failure("Failed to create link token", e.getMessage()));
    }
  }

  @PostMapping("/exchange_public_token")
  public ResponseEntity<ApiResponse<Map<String, Object>>> exchangePublicToken(
      @RequestBody ExchangeTokenRequest request,
      Authentication authentication) {
    try {
      String userId = AuthUtils.requireUserId(authentication);
      PlaidItem item = plaidService.exchangePublicToken(
          userId,
          request.publicToken,
          request.institutionId,
          request.institutionName
      );
      return ResponseEntity.ok(ApiResponse.success(Map.of(
          "itemId", item.getItemId(),
          "institutionName", item.getInstitutionName()
      )));
    } catch (Exception e) {
      return ResponseEntity.badRequest().body(ApiResponse.failure("Failed to exchange token", e.getMessage()));
    }
  }

  @PostMapping("/sync_transactions")
  public ResponseEntity<ApiResponse<Map<String, Object>>> syncTransactions(Authentication authentication) {
    try {
      String userId = AuthUtils.requireUserId(authentication);
      PlaidService.SyncResult result = plaidService.syncTransactions(userId);
      return ResponseEntity.ok(ApiResponse.success(Map.of(
          "added", result.added(),
          "modified", result.modified(),
          "removed", result.removed(),
          "message", "Sync completed successfully"
      )));
    } catch (Exception e) {
      return ResponseEntity.badRequest().body(ApiResponse.failure("Failed to sync transactions", e.getMessage()));
    }
  }

  public static class ExchangeTokenRequest {
    public String publicToken;
    public String institutionId;
    public String institutionName;
  }
}
